addons/udes_stock/models/stock_picking_batch.py

```diff
@@ -1017,21 +1033,45 @@ def reserve_pallet(self, pallet_name):
         Raises a ValidationError if the pallet is already reserved for another
         batch.
         """
+        Picking = self.env['stock.picking']
         PickingBatch = self.env['stock.picking.batch']
 
         self.ensure_one()
 
-        conflicting_batch = PickingBatch.search([
-            ('id', '!=', self.id),
-            ('state', '=', 'in_progress'),
-            ('u_last_reserved_pallet_name', '=', pallet_name),
-        ])
-        if conflicting_batch:
-            raise ValidationError(
-                _("This pallet is already being used for batch %s.") %
-                conflicting_batch[0].name)
+        reserve_pallet_per_picking = self.picking_type_ids.filtered('u_reserve_pallet_per_picking')
+        if reserve_pallet_per_picking:
+            if not picking:
+                raise ValidationError('A picking must be specified if pallets are reserved per picking.')
+        if reserve_pallet_per_picking and picking and not self.picking_ids & picking:
+            return
 
-        self.write({ 'u_last_reserved_pallet_name': pallet_name })
+        if reserve_pallet_per_picking:
+            conflicting_picking = Picking.search([
+                ('u_reserved_pallet', '=', pallet_name),
+                ('state', 'not in', ['draft', 'cancelled', 'done'])
+            ])
+            if conflicting_picking:
+                raise ValidationError(
+                    _("This pallet is already being used for picking %s.") %
+                    conflicting_picking[0].name)
+        else:
+            conflicting_batch = PickingBatch.search([
+                ('id', '!=', self.id),
+                ('state', '=', 'in_progress'),
+                ('u_last_reserved_pallet_name', '=', pallet_name),
+            ])
+            if conflicting_batch:
+                raise ValidationError(
+                    _("This pallet is already being used for batch %s.") %
+                    conflicting_batch[0].name)
+
+        if reserve_pallet_per_picking:
+            # The front end will always add a picking to a batch and then
```

Comment:
Not strictly true, we can create a batch with 4 and assign 4. The thing is that reserve_pallet will receive only one picking as parameter, that is the one we are updating.